datumaro/cli/commands/diff.py

```diff
@@ -2,6 +2,206 @@
 #
 # SPDX-License-Identifier: MIT
 
-# pylint: disable=unused-import
+from enum import Enum, auto
+import argparse
+import json
+import logging as log
+import os
+import os.path as osp
 
-from ..contexts.project import build_diff_parser as build_parser
\ No newline at end of file
+from datumaro.components.errors import ProjectNotFoundError
+from datumaro.components.operations import DistanceComparator, ExactComparator
+from datumaro.util import error_rollback
+from datumaro.util.os_util import rmtree
+from ..util import MultilineFormatter
+from ..util.errors import CliException
+from ..util.project import (generate_next_file_name, load_project,
+    parse_full_revpath)
+from ..contexts.project.diff import DatasetDiffVisualizer
+
+
+class ComparisonMethod(Enum):
+        equality = auto()
+        distance = auto()
+
+eq_default_if = ['id', 'group'] # avoid https://bugs.python.org/issue16399
+
+def build_parser(parser_ctor=argparse.ArgumentParser):
+    parser = parser_ctor(help="Compares two datasets",
+        description="""
+        Compares two datasets. This command has multiple forms:|n
+        1) %(prog)s <revpath>|n
+        2) %(prog)s <revpath> <revpath>|n
+        |n
+        1 - Compares the current project's main target ('project')
+        in the working tree with the specified dataset.|n
+        2 - Compares two specified datasets.|n
+        |n
+        <revpath> - either a dataset path or a revision path. The full
+        syntax is:|n
+        - Dataset paths:|n
+        |s|s- <dataset path>[ :<format> ]|n
+        - Revision paths:|n
+        |s|s- <project path> [ @<rev> ] [ :<target> ]|n
+        |s|s- <rev> [ :<target> ]|n
+        |s|s- <target>|n
```

Comment:
Seems like there is some ambiguity in the syntax. How do you tell apart a `<target>` from a `<rev>` from a `<dataset/project path>`? Perhaps some way to disambiguate should be added?

Reply:
There are 3 or 4 overlapping cases, they are resolved by sequential check of all variants. If any matches - it is accepted, if multiple - they are prioritized as written in the doc, if none - an error is raised. Check `parse_full_revpath` in `datumaro/cli/util/project.py`